bullet/src/Base.hh

```diff
@@ -0,0 +1,214 @@
+  int constraintType;
+  ignition::math::Vector3d axis;
+};
+
+inline btMatrix3x3 convertMat(Eigen::Matrix3d mat)
+{
+  return btMatrix3x3(mat(0, 0), mat(0, 1), mat(0, 2),
+                     mat(1, 0), mat(1, 1), mat(1, 2),
+                     mat(2, 0), mat(2, 1), mat(2, 2));
+}
+
+inline btVector3 convertVec(Eigen::Vector3d vec)
+{
+  return btVector3(vec(0), vec(1), vec(2));
+}
+
+inline Eigen::Matrix3d convert(btMatrix3x3 mat)
+{
+  Eigen::Matrix3d val;
+  val << mat[0][0], mat[0][1], mat[0][2],
+         mat[1][0], mat[1][1], mat[1][2],
+         mat[2][0], mat[2][1], mat[2][2];
+  return val;
+}
+
+inline Eigen::Vector3d convert(btVector3 vec)
+{
+  Eigen::Vector3d val;
+  val << vec[0], vec[1], vec[2];
+  return val;
+}
+
+class Base : public Implements3d<FeatureList<Feature>>
+{
+  public: std::size_t entityCount = 0;
+  public: std::size_t collisionGroup = 1;
+
+  public: inline std::size_t GetNextEntity()
+  {
+    return entityCount++;
+  }
+
+  public: inline std::size_t GetNextCollisionGroup()
+  {
+    return collisionGroup++;
+  }
+
+  public: inline Identity InitiateEngine(std::size_t /*_engineID*/) override
+  {
+    const auto id = this->GetNextEntity();
+    assert(id == 0);
+
+    return this->GenerateIdentity(0);
+  }
+
+  public: inline Identity AddWorld(WorldInfo _worldInfo)
+  {
+    const auto id = this->GetNextEntity();
+    this->worlds[id] = std::make_shared<WorldInfo>(_worldInfo);
+    return this->GenerateIdentity(id, this->worlds.at(id));
+  }
+
+  public: inline Identity AddModel(ModelInfo _modelInfo)
+  {
+    const auto id = this->GetNextEntity();
+    this->models[id] = std::make_shared<ModelInfo>(_modelInfo);
+
+    // A collision mask for each different model
+    this->collisionGroups[id] = 1 << GetNextCollisionGroup();
+
+    return this->GenerateIdentity(id, this->models.at(id));
+  }
+
+  public: inline Identity AddLink(LinkInfo _linkInfo)
+  {
+    const auto id = this->GetNextEntity();
+    this->links[id] = std::make_shared<LinkInfo>(_linkInfo);
+
+    auto model = this->models.at(_linkInfo.model);
+    model->links.push_back(id);
+
+    return this->GenerateIdentity(id, this->links.at(id));
+  }
+  public: inline Identity AddCollision(CollisionInfo _collisionInfo)
+  {
+   const auto id = this->GetNextEntity();
+   this->collisions[id] = std::make_shared<CollisionInfo>(_collisionInfo);
+
+   return this->GenerateIdentity(id, this->collisions.at(id));
+  }
+
+  public: inline Identity AddJoint(JointInfo _jointInfo)
+  {
+    const auto id = this->GetNextEntity();
+    this->joints[id] = std::make_shared<JointInfo>(_jointInfo);
+
+    return this->GenerateIdentity(id, this->joints.at(id));
+  }
+
+  public: using WorldInfoPtr = std::shared_ptr<WorldInfo>;
+  public: using ModelInfoPtr = std::shared_ptr<ModelInfo>;
+  public: using LinkInfoPtr  = std::shared_ptr<LinkInfo>;
+  public: using CollisionInfoPtr = std::shared_ptr<CollisionInfo>;
+  public: using JointInfoPtr  = std::shared_ptr<JointInfo>;
```

Comment:
Any reason for `shared` vs `unique` here?

Reply:
Those in particular can't be changed because the `GenerateIdentity` templated function from ignition takes a shared pointer to generate new identities.